widgetssdk/src/main/java/com/glia/widgets/chat/domain/AppendNewMessageUseCase.kt

```diff
@@ -0,0 +1,181 @@
+import com.glia.androidsdk.chat.SystemMessage
+import com.glia.androidsdk.chat.VisitorMessage
+import com.glia.widgets.chat.domain.gva.IsGvaUseCase
+import com.glia.widgets.chat.model.ChatItem
+import com.glia.widgets.chat.model.OperatorChatItem
+import com.glia.widgets.chat.model.VisitorChatItem
+import com.glia.widgets.chat.model.VisitorMessageItem
+import com.glia.widgets.core.engagement.domain.model.ChatMessageInternal
+import com.glia.widgets.helper.Logger
+import com.glia.widgets.helper.TAG
+import com.glia.widgets.helper.asSingleChoice
+
+
+internal class AppendNewChatMessageUseCase(
+    private val appendNewOperatorMessageUseCase: AppendNewOperatorMessageUseCase,
+    private val appendNewVisitorMessageUseCase: AppendNewVisitorMessageUseCase,
+    private val appendSystemMessageItemUseCase: AppendSystemMessageItemUseCase
+) {
+    operator fun invoke(chatMessagesState: ChatMessagesState, chatMessageInternal: ChatMessageInternal) {
+        when (val message = chatMessageInternal.chatMessage) {
+            is VisitorMessage -> {
+                appendNewVisitorMessageUseCase(chatMessagesState, chatMessageInternal)
+                chatMessagesState.resetOperator()
+            }
+
+            is OperatorMessage -> appendNewOperatorMessageUseCase(chatMessagesState, chatMessageInternal)
+
+            is SystemMessage -> {
+                appendSystemMessageItemUseCase(chatMessagesState.chatItems, message)
+                chatMessagesState.resetOperator()
+            }
+
+            else -> Logger.d(TAG, "Unexpected type of message received -> $message")
+        }
+    }
+}
+
+internal class AppendNewOperatorMessageUseCase(
+    private val isGvaUseCase: IsGvaUseCase,
+    private val customCardAdapterTypeUseCase: CustomCardAdapterTypeUseCase,
+    private val appendGvaMessageItemUseCase: AppendGvaMessageItemUseCase,
+    private val appendHistoryCustomCardItemUseCase: AppendHistoryCustomCardItemUseCase,
+    private val appendNewResponseCardOrTextItemUseCase: AppendNewResponseCardOrTextItemUseCase
+) {
+    operator fun invoke(chatMessagesState: ChatMessagesState, chatMessageInternal: ChatMessageInternal) {
+        val itemsCount = chatMessagesState.chatItems.count()
+        val message: OperatorMessage = chatMessageInternal.chatMessage as OperatorMessage
+        when {
+            isGvaUseCase(message) -> appendGvaMessageItemUseCase(chatMessagesState.chatItems, chatMessageInternal)
+            customCardAdapterTypeUseCase(message) != null -> appendHistoryCustomCardItemUseCase(
+                chatMessagesState.chatItems,
+                message,
+                customCardAdapterTypeUseCase(message)!!
+            )
+
+            else -> appendNewResponseCardOrTextItemUseCase(chatMessagesState.chatItems, chatMessageInternal)
+        }
+
+        chatMessagesState.apply { addedMessagesCount = chatItems.count() - itemsCount }
+
+
+        val lastChatHead = chatMessagesState.lastChatHeadItem
```

Comment:
It is a little hard to understand what the chatHeadItem is. Maybe rename it to something better